docs/operators/deduper.md

```diff
@@ -0,0 +1,523 @@
+operator within some time after they are generated. After this time, the
+tuples may be considered as stale or obsolete.
+
+In such cases, the Deduper chooses to consider these tuples
+expired and takes no action but to
+separate out these tuples on a different port in order to be processed
+by some other operator or offline analysis.
+
+In order to create a criterion for discarding such tuples, we
+introduce an Expiry Key. Looking at the value of the Expiry Key in each
+tuple, we can decide whether or not to discard this tuple as
+expired.
+
+The expiry key that we consider in Time Based Dedup is
+time. This usually works with
+another parameter called Expiry Period defined next.
+
+### Expiry Period
+
+The Expiry Period is the value supplied by the user which decides
+when a particular tuple expires.
+
+### Time Points
+
+For every dataset that the deduper processes, it maintains a set
+of time points:
+
+1.  *Latest Point* - This is the maximum
+    time point observed in all the processed tuples.
+2.  *Expiry Point* - This is given by:
+    `Expiry Point = Latest Point - Expiry`
+    Period
+
+These points help the deduper to make decisions related to expiry
+of a tuple.
+
+### Example - Expiry
+
+|Tuple Id|Expiry Key (Expiry Period = 10)|Latest Point|Expiry Point|Decision for Tuple|
+|:------:|:-----------------------------:|:----------:|:----------:|:----------------:|
+|1|10|10|1|Not Expired|
+|2|20|20|11|Not Expired|
+|3|25|25|16|Not Expired|
+|4|40|40|31|Not Expired|
+|5|21|40|31|Expired|
+|6|35|40|31|Not Expired|
+|7|45|45|36|Not Expired|
+|8|57|57|48|Not Expired|
+
+
+### Time Buckets (A component of Managed State)
+
+One of the requirements of the Deduper is to store all the unique
+tuples (actually, just the keys of tuples). Keeping an ever growing
+cache in memory is not scalable. So what we need is a limited cache
+backed by a persistent store. Now to reduce cache misses, we load a
+chunk of data (called Buckets), together into memory. Buckets help
+narrow down the search of duplicates for incoming tuples. A Bucket is an
+abstraction for a collection of tuples all of which share a common hash
+value based on some hash function or a range of time, for example: a
+bucket of data for5 contiguous minutes. A Bucket has a span property called Bucket Span.
+
+### Bucket Span
+
+Bucket span is simply the range of the domain
+that is covered by the Bucket. This span is specified in
+the domain of the Expiry key. If the Expiry
+Key is time,  then the Bucket span
+will be specified in seconds. It is
+only defined in case tuples have an Expiry Key.
+
+### Number of Buckets
+
+The number of buckets can be given by - Num Buckets = Expiry
+Period / Bucket Span.
+
+This is because at any point of time, we need only store Expiry
+Period worth of data.
+
+### Example - Buckets
+
+![](images/deduper/image01.png)
+
+Assumptions
+------------------------
+
+### Assumption 1 <a name="Assumption1"></a>
+
```

Comment:
Suggest adding a descriptive version of the assumption, e.g. Any two tuples with the same dedup key are assumed to have the same expiry key as well.